### PR TITLE
Third party libs: improve documentation + introduce new CMake functions

### PR DESCRIPTION
**Pull request by sgiraudot**
## Summary of Changes

This PR improves how third party libraries for Point Set Processing related packages are handled:
1. Generalize the `CGAL_target_use_[lib](<target>)` introduced with TBB. This allows to be cleaner in terms of inclusion (use `target_include_directories()` instead of `include_directories()` to avoid polluting other targets with useless includes), handle the macros (like `CGAL_LINKED_WITH_LASLIB`) and linking. For Eigen, it may look less user-friendly because you now have to specify which target use Eigen, but it seems much much cleaner to me: I found out that some targets were not in the `if(EIGEN3_FOUND)` section whereas they _did_ require Eigen. It only worked before because the Eigen files were included globally.
2. Document missing libraries in the advanced section "_Summary of CGAL's Configuration Variables_" of the manual
3. Document the `CGAL_target_use_[lib](<target>)` functions in the "_Optional Third Party_" section. It's debatable, but I think these should be documented as they make the life of the users much simpler (so far, they have to look in the `CMakeLists.txt` of the examples to have an idea of what's going on).  I put the documentation in this section because it's the section where we usually redirect users when we document that a feature depends on a third party library, so it should be the most easy to find.

I'm putting the _Under discussion_ tag as I think this should be discussed a bit before being integrated (or not). The scripts look now much cleaner to me (especially if you look at the Classification example script, it was barely readable before, and now it's quite okay), but it might not be the case for everybody.

## Release Management

* Affected package(s): pretty much every package that used Eigen + all PSP related packages
* Issue(s) solved (if any): fix https://github.com/CGAL/cgal/issues/4215

### CODE REVIEW

**Comment by sgiraudot** [comment]
As no one complained about this feature (and @lrineau and I discussed it AFK and agreed on it), I propose to test it for integration.

**Comment by afabri** [comment]
I am not sure if we want to  add CMake functions.  Real user have to integrate CGAL components into their applications so it should be "raw CMake" that they can copy paste (the non-CMake scenario is yet another story).

File: Polyhedron/demo/Polyhedron/Plugins/Mesh_3/CMakeLists.txt [hunk 1]
```diff
@@ -51,7 +51,14 @@ polyhedron_demo_plugin(mesh_3_optimization_plugin Optimization_plugin
...
 
-
+# Use Eigen
+ find_package(Eigen3 3.1.0 REQUIRED) #(3.1.0 or greater)
+ 
+ if (EIGEN3_FOUND)
+   CGAL_target_use_Eigen(mesh_3_optimization_plugin)
+ else() #eigen
+   message(STATUS "The Mesh_3_optimization_plugin requires Eigen, which was not found, and will use a deprecated class to replace it. Warnings are to be expected.")
+ endif()#eigen
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Mesh_3/CMakeLists.txt]: With `REQUIRED` in the `find_package`, the `if(EIGEN3_FOUND)` is useless. Are you sure you wanted that `REQUIRED`?

**Comment by maxGimeno** [comment]
@sgiraudot maybe this PR needs some updates with master, to add the missing Eigen managemen, at least in mesh smoothing

**Comment by sgiraudot** [comment]
Sure, I'll take care of it.

**Comment by maxGimeno** [comment]
thx !

**Comment by sgiraudot** [comment]
I updated with master and run the full testsuite on my machine, but I did not find any missing Eigen management…

What I did find was that a test in `Number_types` called `eigen.cpp` was actually never compiled with Eigen and only displayed `Eigen is not configured!`. And of course, when I actually included Eigen… it did not compile well. I had to remove to tests because apparently, Eigen does not compute determinants with integer matrices. It should work now.

**Comment by maxGimeno** [comment]
Thank you ! We'll see that tomorrow

**Comment by maxGimeno** [comment]
[Errors](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-72.shtml#Number_types) 
plus [travis errors](https://travis-ci.org/CGAL/cgal/jobs/648447615#L3703), which are actually the same.

File: Polygonal_surface_reconstruction/test/Polygonal_surface_reconstruction/CMakeLists.txt [hunk 2]
```diff
@@ -24,58 +24,38 @@ include( ${CGAL_USE_FILE} )
...
-    include_directories( BEFORE ${GLPK_INCLUDE_DIR} )
-
-    target_link_libraries( polygonal_surface_reconstruction_test PRIVATE ${GLPK_LIBRARIES} )
-
-    target_compile_definitions(polygonal_surface_reconstruction_test PRIVATE -DCGAL_USE_GLPK)
-
-    message("GLPK found and used")
-
+  if ( NOT GLPK_FOUND )
+    message( STATUS "NOTICE: This project requires either SCIP or GLPK, and will not be compiled.")
```
Comment (maxGimeno) [inline Polygonal_surface_reconstruction/test/Polygonal_surface_reconstruction/CMakeLists.txt]:
```suggestion
    message( STATUS "NOTICE : This project requires either SCIP or GLPK, and will not be compiled.")
```

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-88.shtml

**Comment by lrineau** [comment]
> Successfully tested in [cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-88.shtml](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-88.shtml)

This PR was **not** tested in [CGAL-5.1-Ic-88.shtml](https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-88.shtml)!

File: Number_types/test/Number_types/eigen.cpp [hunk 3]
```diff
@@ -59,7 +59,7 @@ int main(){
   //check<CGAL::Lazy_exact_nt<CGAL::Quotient<CGAL::MP_Float> > >();
   check<CGAL::Sqrt_extension<double,double> >();
 #ifdef CGAL_USE_GMP
-  check<CGAL::Gmpz>();
+//  check<CGAL::Gmpz>();
```
Comment (maxGimeno) [inline Number_types/test/Number_types/eigen.cpp]: ???

**Comment by maxGimeno** [comment]
[Errors](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-95/Number_types/TestReport_afabri_x64_Cygwin-Windows10_MSVC2015-Debug-64bits.gz) on MSVC 2015

File: Number_types/test/Number_types/eigen.cpp (hunk 3 above)
Comment (sgiraudot) [reply Number_types/test/Number_types/eigen.cpp]: Eigen does not work on integer types. As this file was never tested before, this never raised an error but it does not work.

I could have removed it completely but I figured I'd just comment it (similarly to line 59) as someone might think the test is just missing and no deactivated voluntarily.

**Comment by maxGimeno** [comment]
That last commit is not correct :/

**Comment by sgiraudot** [comment]
It should be fixed now.

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-100.shtml

**Pull request merged**